File: 0079-word-search/0079-word-search.java
class Solution {
    private boolean found = false;
    
    public boolean exist(char[][] board, String word) {
    for (int row = 0; row < board.length; row++)
      for (int col = 0; col < board[0].length; col++) {
        dfs(row, col, word, board);
        if (found)
            return true;
      }
        return false;
    }
    
    private void dfs(int r, int c, String str, char[][] board) {
        char ch;
        if (str.equals("")) {
            found = true;
            return;            
        }
        
        if (r >= board.length || r < 0 || c < 0 || c >= board[0].length || board[r][c] != str.charAt(0))
            return;
        else {
            ch = board[r][c];
            board[r][c] = '0';
            dfs(r + 1, c, str.substring(1, str.length()), board);
            dfs(r - 1, c, str.substring(1, str.length()), board);
            dfs(r, c + 1, str.substring(1, str.length()), board);
            dfs(r, c - 1, str.substring(1, str.length()), board);
        }
        board[r][c] = ch;
    }
}

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        def dfs(i, j, k):
            if k >= len(word):
                return True
            if i < 0 or j < 0 or i >= len(board) or j >= len(board[0]):
                return False
            if board[i][j] != word[k]:
                return False
            temp = board[i][j]
            board[i][j] = '0'
            res = dfs(i - 1, j, k + 1) or dfs(i + 1, j, k + 1) or dfs(i, j + 1, k + 1) or dfs(i, j - 1, k + 1)
            board[i][j] = temp
            return res
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                if dfs(i, j, 0):
                    return True
        
        return False